### PR TITLE
US #657060 / #671115 - [ADO] Have ALM Lab Mgmt run with Automatch option for functional tests

### PR DESCRIPTION
**Pull request by dorin7bogdan**

### CODE REVIEW

File: PSModule/AlmLabMgmtClient/SDK/Auth/RestAuthenticator.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            {
+                { HttpRequestHeader.ContentType, C.APP_XML },
+                { HttpRequestHeader.Accept, C.APP_XML }
+            };
+
+            // issue a post request so that cookies relevant to the QC Session will be added to the RestClient
+            Response res = await client.HttpPost(
+                client.ServerUrl.AppendSuffix(SESSION_ENDPOINT), 
+                headers,
+                $"<session-parameters><client-type>{clientType}</client-type></session-parameters>");
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Auth/RestAuthenticator.cs:63]: Is that possible the `clientType` may contain `<`, `>` or other invalid XML characters? If so, may need to use `CDATA` to make sure the string value shall not break the XML rules.

File: PSModule/AlmLabMgmtClient/SDK/Handler/EventLogHandler.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+            }
+
+            return ok;
+        }
+
+        private bool IsNew(IDictionary<string, string> currEntity)
+        {
+            bool isNew = false;
+            int currEvent = int.Parse(currEntity[ID]);
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Handler/EventLogHandler.cs:59]: 1. The `ID` may not exist in dictionary.
2. `int.Parse` may throw exception, use `int.TryParse` instead.

File: PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            try
+            {
+                string xml = response.ToString();
+                string endTime = Xml.GetAttributeValue(xml, END_TIME);
+                if (!endTime.IsNullOrWhiteSpace())
+                {
+                    string startTime = Xml.GetAttributeValue(xml, START_TIME);
+                    string currentRunState = Xml.GetAttributeValue(xml, STATE);
+                    _logger.LogInfo($"Timeslot {_timeslotId} is [{currentRunState}].\nRun start time: [{startTime}], Run end time: [{endTime}]");
+                    ret = true;
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs:68]: Do we need to check if the `startTime` and `currentRunState` is not empty and valid before setting `ret` to `true`?

File: PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        protected override bool LogRunEntityResults(Response response)
+        {
+            bool ok = false;
+            try
+            {
+                string xml = response.ToString();
+                string state = Xml.GetAttributeValue(xml, STATE);
+                string completedSuccessfully = Xml.GetAttributeValue(xml, COMPLETED_SUCCESSFULLY);
+                _logger.LogInfo($"Run state of {_runId}: {state}, Completed successfully: {completedSuccessfully}");
+                ok = true;
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs:93]: Before setting `ok` to `true`, do we need check if `completedSuccessfully` is one of the values that we treat it as success?

File: PSModule/AlmLabMgmtClient/SDK/Handler/PollHandler.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    public abstract class PollHandler : HandlerBase
+    {
+        private readonly int _interval = 5000; // millisecond
+        protected PollHandler(IClient client, string entityId) : base(client, entityId)
+        {
+        }
+
+        protected PollHandler(IClient client, string entityId, int interval) : base(client, entityId)
+        {
+            _interval = interval;
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Handler/PollHandler.cs:17]: Need to check if the `interval` is valid.
https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.sleep?view=net-5.0#System_Threading_Thread_Sleep_System_Int32_

File: PSModule/AlmLabMgmtClient/SDK/Request/PostRequestBase.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+                    ResourceAccessLevel.PROTECTED,
+                    logRequestUrl);
+        }
+
+        private string GetXmlData()
+        {
+            StringBuilder builder = new StringBuilder("<Entity><Fields>");
+            foreach (KeyValuePair<string, string> pair in DataFields)
+            {
+                builder.Append($"<Field Name=\"{pair.Key}\"><Value>{pair.Value}</Value></Field>");
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Request/PostRequestBase.cs:38]: 1. Here `pair.Value` may contains invalid XML characters. Use `CDATA` section to wrap the value.
2. The `pair.Key` may also contains invalid XML characters that may damage entire XML, may need to check before constructing the XML string.

File: PSModule/AlmLabMgmtClient/SDK/Util/AlmRunReportUrlBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                // result url will be NA (in case of failure like getting ALM version, convert ALM version to number)
+                Console.WriteLine(e.Message);
+            }
+            return url;
+        }
+
+        public async Task<bool> IsNewReport(IClient client)
+        {
+            AlmVersion version = await GetAlmVersion(client);
```
Comment (soleverxu) [inline PSModule/AlmLabMgmtClient/SDK/Util/AlmRunReportUrlBuilder.cs:40]: The `GetAlmVersion` method may throw exception, here need to catch and return `false`.

**Review by soleverxu — COMMENTED** [review]
See comments inline

**Review by soleverxu — CHANGES_REQUESTED** [review]
May need change code

File: PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs (hunk 3 above)
Comment (dorin7bogdan) [reply PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs:68]: The idea is very good, but currently we don't know if we will get always the same time format across all ALM server versions.  If the received value does not match the expected formats the code will enter an infinite loop.
If the received value is not a timestamp, the code will stop the polling and exit the loop, which is not the best behavior, but from the printed output the user can observe and report the issue.

Comment (soleverxu) [reply PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs:68]: Got you. Thanks.

File: PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs (hunk 4 above)
Comment (dorin7bogdan) [reply PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs:93]: The function LogRunEntityResults should return true if the response is valid (contains the expected attributes, regardless their value). The real test result will be queried and printed later using LabPublisher.cs.

Comment (soleverxu) [reply PSModule/AlmLabMgmtClient/SDK/Handler/LabPollHandler.cs:93]: You mentioned **valid**, however, the variable `state` and `completedSuccessfully` are never checked. For example, if `completedSuccessfully` is empty?

File: PSModule/AlmLabMgmtClient/SDK/Auth/RestAuthenticator.cs (hunk 1 above)
Comment (dorin7bogdan) [reply PSModule/AlmLabMgmtClient/SDK/Auth/RestAuthenticator.cs:63]: Invalid values are allowed at least from Azure UI (in most cases this field is left empty). 
For this issue I would ask the QA to reproduce and open a defect, then we can fix and validate it specifically.
What do you think?

Comment (soleverxu) [reply PSModule/AlmLabMgmtClient/SDK/Auth/RestAuthenticator.cs:63]: Sure. No problem.

File: PSModule/AlmLabMgmtClient/SDK/Request/PostRequestBase.cs (hunk 6 above)
Comment (dorin7bogdan) [reply PSModule/AlmLabMgmtClient/SDK/Request/PostRequestBase.cs:38]: 1. The `pair.Value` might contain invalid XML characters. For this issue I would ask the QA to reproduce and open a defect, then we can fix and validate it specifically.
What do you think?
2. The `pair.Key` contains always predefined constants with valid characters, at least at this moment.

Comment (soleverxu) [reply PSModule/AlmLabMgmtClient/SDK/Request/PostRequestBase.cs:38]: Sure. No problem.

**Review by soleverxu — APPROVED** [review]
Two XML parse issues may break functionalities, need more tests and if issues are found, need to fix asap.

**Pull request merged**